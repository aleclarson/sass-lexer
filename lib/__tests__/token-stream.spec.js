// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

/* eslint-env jest */

const tokenize = require('..')

describe('#peek', () => {
  it('returns the current token', () => {
    let t = tokenize('hello')
    expect(t.peek()).toMatchSnapshot()
  })
  it('returns the current token with an offset', () => {
    let t = tokenize('hello world')
    expect(t.peek(1)).toMatchSnapshot()
  })
})

describe('#next', () => {
  it('consumes returns and the next token', () => {
    let t = tokenize('hello world')
    expect(t.next()).toMatchSnapshot()
    expect(t.peek()).toMatchSnapshot()
  })
  describe('tokens', () => {
    describe('space', () => {
      it('single space', () => {
        let t = tokenize(' ')
        expect(t.next()).toMatchSnapshot()
      })
      it('multiple spaces', () => {
        let t = tokenize('    hello')
        expect(t.next()).toMatchSnapshot()
      })
      it('whitespace characters', () => {
        let t = tokenize('\n\n\t  hello')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('comment', () => {
      it('single comment', () => {
        let t = tokenize('// Hello\nWorld')
        expect(t.next()).toMatchSnapshot()
      })
      it('single comment', () => {
        let t = tokenize('/** Hello World */')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('number', () => {
      it('integer', () => {
        let t = tokenize('3')
        expect(t.next()).toMatchSnapshot()
      })
      it('float', () => {
        let t = tokenize('3.0')
        expect(t.next()).toMatchSnapshot()
      })
      it('float (leading decimal)', () => {
        let t = tokenize('.3')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('hex', () => {
      it('6 digit lowercase', () => {
        let t = tokenize('#ff0099')
        expect(t.next()).toMatchSnapshot()
      })
      it('6 digit uppercase', () => {
        let t = tokenize('#FF0099')
        expect(t.next()).toMatchSnapshot()
      })
      it('3 digit lowercase', () => {
        let t = tokenize('#ff0')
        expect(t.next()).toMatchSnapshot()
      })
      it('3 digit uppercase', () => {
        let t = tokenize('#FF0')
        expect(t.next()).toMatchSnapshot()
      })
      it('3 digit (trailing invalid)', () => {
        let t = tokenize('#FF0;')
        expect(t.next()).toMatchSnapshot()
      })
      it('6 digit numbers', () => {
        let t = tokenize('#000000')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('atrule', () => {
      it('works', () => {
        let t = tokenize('@mixin')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('punctuation', () => {
      it('{', () => {
        let t = tokenize('{')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('operator', () => {
      it('+', () => {
        let t = tokenize('+')
        expect(t.next()).toMatchSnapshot()
      })
      it('repeatable', () => {
        let t = tokenize('&&')
        expect(t.next()).toMatchSnapshot()
      })
      it('non-repeatable', () => {
        let t = tokenize('++')
        expect(t.next()).toMatchSnapshot()
      })
      it('repeatable followed by non-repeatable', () => {
        let t = tokenize('&++')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('identifier', () => {
      it('checks for valid starting characters', () => {
        let t = tokenize('_hello world')
        expect(t.next()).toMatchSnapshot()
      })
      it('ignores invalid starting characters', () => {
        let t = tokenize('0hello world')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('string', () => {
      it('single quotes', () => {
        let t = tokenize('\'hello\'')
        expect(t.next()).toMatchSnapshot()
      })
      it('double quotes', () => {
        let t = tokenize('"hello"')
        expect(t.next()).toMatchSnapshot()
      })
      it('escaped characters', () => {
        let t = tokenize('"hello \\"world\\""')
        expect(t.next()).toMatchSnapshot()
      })
      it('preserves escape characters', () => {
        let t = createTokenStream(createInputStream('token(\'\'+myVar+\'font(\\\'world\\\')\')'))
        expect(t.all()).toMatchSnapshot()
      })
    })
    describe('variable', () => {
      it('works', () => {
        let t = tokenize('$size')
        expect(t.next()).toMatchSnapshot()
      })
    })
    describe('sink', () => {
      it('1', () => {
        let t = createTokenStream(createInputStream('($var)'))
        expect(t.all()).toMatchSnapshot()
      })
      it('2', () => {
        let t = createTokenStream(createInputStream('// ($var)\n@mixin myMixin'))
        expect(t.all()).toMatchSnapshot()
      })
    })
  })
})

describe('#eof', () => {
  it('returns false if there are more tokens', () => {
    let t = tokenize('hello')
    expect(t.eof()).toEqual(false)
  })
  it('returns true if there are no more tokens', () => {
    let t = tokenize('hello world')
    expect(t.eof()).toEqual(false)
    t.next()
    t.next()
    t.next()
    expect(t.eof()).toEqual(true)
  })
})

describe('#err', () => {
  it('throws an error', () => {
    let t = tokenize('hello world')
    t.next()
    t.next()
    expect(() => {
      t.err('Whoops')
    }).toThrow(/Whoops \(1:6\)/)
  })
})
