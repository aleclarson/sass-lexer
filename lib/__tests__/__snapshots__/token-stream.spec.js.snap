// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#next consumes returns and the next token 1`] = `
Array [
  "identifier",
  "hello",
  1,
  0,
]
`;

exports[`#next consumes returns and the next token 2`] = `
Array [
  "space",
  " ",
  1,
  5,
]
`;

exports[`#next tokens atrule works 1`] = `
Array [
  "atrule",
  "mixin",
  1,
  0,
]
`;

exports[`#next tokens comment single comment 1`] = `
Array [
  "comment",
  " Hello",
  1,
  0,
]
`;

exports[`#next tokens comment single comment 2`] = `
Array [
  "comment",
  "* Hello World ",
  1,
  0,
]
`;

exports[`#next tokens hex 3 digit (trailing invalid) 1`] = `
Array [
  "color_hex",
  "FF0",
  1,
  0,
]
`;

exports[`#next tokens hex 3 digit lowercase 1`] = `
Array [
  "color_hex",
  "ff0",
  1,
  0,
]
`;

exports[`#next tokens hex 3 digit uppercase 1`] = `
Array [
  "color_hex",
  "FF0",
  1,
  0,
]
`;

exports[`#next tokens hex 6 digit lowercase 1`] = `
Array [
  "color_hex",
  "ff0099",
  1,
  0,
]
`;

exports[`#next tokens hex 6 digit numbers 1`] = `
Array [
  "color_hex",
  "000000",
  1,
  0,
]
`;

exports[`#next tokens hex 6 digit uppercase 1`] = `
Array [
  "color_hex",
  "FF0099",
  1,
  0,
]
`;

exports[`#next tokens identifier checks for valid starting characters 1`] = `
Array [
  "identifier",
  "_hello",
  1,
  0,
]
`;

exports[`#next tokens identifier ignores invalid starting characters 1`] = `
Array [
  "number",
  "0",
  1,
  0,
]
`;

exports[`#next tokens number float (leading decimal) 1`] = `
Array [
  "number",
  ".3",
  1,
  0,
]
`;

exports[`#next tokens number float 1`] = `
Array [
  "number",
  "3.0",
  1,
  0,
]
`;

exports[`#next tokens number integer 1`] = `
Array [
  "number",
  "3",
  1,
  0,
]
`;

exports[`#next tokens operator + 1`] = `
Array [
  "operator",
  "+",
  1,
  0,
]
`;

exports[`#next tokens operator non-repeatable 1`] = `
Array [
  "operator",
  "+",
  1,
  0,
]
`;

exports[`#next tokens operator repeatable 1`] = `
Array [
  "operator",
  "&&",
  1,
  0,
]
`;

exports[`#next tokens operator repeatable followed by non-repeatable 1`] = `
Array [
  "operator",
  "&",
  1,
  0,
]
`;

exports[`#next tokens punctuation { 1`] = `
Array [
  "punctuation",
  "{",
  1,
  0,
]
`;

exports[`#next tokens sink 1 1`] = `
Array [
  Array [
    "punctuation",
    "(",
    1,
    0,
  ],
  Array [
    "variable",
    "var",
    1,
    1,
  ],
  Array [
    "punctuation",
    ")",
    1,
    5,
  ],
]
`;

exports[`#next tokens sink 2 1`] = `
Array [
  Array [
    "comment",
    " ($var)",
    1,
    0,
  ],
  Array [
    "space",
    "
",
    1,
    9,
  ],
  Array [
    "atrule",
    "mixin",
    2,
    0,
  ],
  Array [
    "space",
    " ",
    2,
    6,
  ],
  Array [
    "identifier",
    "myMixin",
    2,
    7,
  ],
]
`;

exports[`#next tokens space multiple spaces 1`] = `
Array [
  "space",
  "    ",
  1,
  0,
]
`;

exports[`#next tokens space single space 1`] = `
Array [
  "space",
  " ",
  1,
  0,
]
`;

exports[`#next tokens space whitespace characters 1`] = `
Array [
  "space",
  "

	  ",
  1,
  0,
]
`;

exports[`#next tokens string double quotes 1`] = `
Array [
  "string",
  "hello",
  1,
  0,
]
`;

exports[`#next tokens string escaped characters 1`] = `
Array [
  "string",
  "hello \\\\\\"world\\\\\\"",
  1,
  0,
]
`;

exports[`#next tokens string preserves escape characters 1`] = `
Array [
  Array [
    "identifier",
    "token",
    1,
    0,
  ],
  Array [
    "punctuation",
    "(",
    1,
    5,
  ],
  Array [
    "string",
    "",
    1,
    6,
  ],
  Array [
    "operator",
    "+",
    1,
    8,
  ],
  Array [
    "identifier",
    "myVar",
    1,
    9,
  ],
  Array [
    "operator",
    "+",
    1,
    14,
  ],
  Array [
    "string",
    "font(\\\\'world\\\\')",
    1,
    15,
  ],
  Array [
    "punctuation",
    ")",
    1,
    32,
  ],
]
`;

exports[`#next tokens string single quotes 1`] = `
Array [
  "string",
  "hello",
  1,
  0,
]
`;

exports[`#next tokens variable works 1`] = `
Array [
  "variable",
  "size",
  1,
  0,
]
`;

exports[`#peek returns the current token 1`] = `
Array [
  "identifier",
  "hello",
  1,
  0,
]
`;

exports[`#peek returns the current token with an offset 1`] = `
Array [
  "space",
  " ",
  1,
  5,
]
`;
